src/main/scala/mesosphere/marathon/core/storage/PersistenceStore.scala

```diff
@@ -0,0 +1,254 @@
+package mesosphere.marathon.core.storage
+
+import akka.http.scaladsl.marshalling.{ Marshal, Marshaller }
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.{ Done, NotUsed }
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.util.CallerThreadExecutionContext
+
+import scala.async.Async._
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+/**
+  * Resolver for Marathon Internal Persistence IDs (by `Key Type`, `Value Type` and `Serialized Type`)
+  * For example, Applications may be identified by the string "/apps/abc" and stored internally at
+  * "/apps/a/a58ec17d-735d-4c3f-9aa8-d44c764aa31b"
+  *
+  * This IdResolver must be provided for all storage engines, e.g.
+  *
+  * {{{
+  *   class A(val id: String) {}
+  *   object A {
+  *     implicit val zkIdResolver = new IdResolver[String, A, ByteString] = {
+  *       def toStorageId(id: String): String = id
+  *       def fromStorageId(id: String): String = id
+  *   }
+  * }}}
+  *
+  * @tparam Id The marathon type of the ID for the given Value type
+  * @tparam K The persisted type of the ID
+  * @tparam V The value type being stored
+  * @tparam Serialized The serialized type of 'V' for the given [[PersistenceStore]].
+  */
+trait IdResolver[Id, K, +V, +Serialized] {
+  /**
+    * Translate the marathon id into the given persisted format
+    */
+  def toStorageId(id: Id): K
+
+  /**
+    * Translate from the persisted format to the marathon id.
+    */
+  def fromStorageId(key: K): Id
+}
+
+/**
+  * Generic Persistence Store with flexible storage backends using Akka Marshalling Infrastructure.
+  *
+  * == Providing Serialization for a given class ==
+  *  - Provide an [[IdResolver]] for your class for the supported [[PersistenceStore]]s
+  *  - Provide a Marshaller and Unmarshaller for
+  *    your class and the Serialized form for the supported [[PersistenceStore]]s.
+  *  - For example, given a class 'A', a K set to [[mesosphere.marathon.core.storage.impl.zk.ZkId]]
+  *    and Serialized as [[mesosphere.marathon.core.storage.impl.zk.ZkSerialized]],
+  *    the following implicits should be sufficient
+  * {{{
+  *   case class A(id: Int, name: String)
+  *   object A {
+  *     implicit val zkIdResolver = new IdResolver[Int, ZkId, A, ZkSerialized] {
+  *       def toStorageId(id: Int): ZkId =
+  *         ZkId("/A/" + id.toString) // note: scaladoc bug where string interpolation fails
+  *       def fromStorageId(key: ZkId): Int = key.value.replaceAll("/A/", "").toInt
+  *     }
+  *     implicit val zkMarshaller = Marshaller[A, ZkSerialized] =
+  *       Marshaller.opaque { (a: A) =>
+  *         val bytes = ByteString.newBuilder
```

Comment:
We used protobufs in the past to marshal objects to byte strings. Are you suggesting to use plain akka primitives here, or is this just an example? 
I think it would be good to use the Marshaller/Unmarshaller traits from akka but use protobuf as implementation.